src/compiler/commandLineParser.ts

```diff
@@ -633,10 +637,22 @@ namespace ts {
                 const supportedExtensions = getSupportedExtensions(options);
                 Debug.assert(indexOf(supportedExtensions, ".ts") < indexOf(supportedExtensions, ".d.ts"), "Changed priority of extensions to pick");
 
+                const potentialFiles: string[] = [];
+                if (host.readDirectoryWithMultipleExtensions) {
+                    addRange(potentialFiles, host.readDirectoryWithMultipleExtensions(basePath, supportedExtensions, exclude));
+                }
+                else {
+                    for (const extension of supportedExtensions) {
+                        addRange(potentialFiles, host.readDirectory(basePath, extension, exclude));
+                    }
+                }
                 // Get files of supported extensions in their order of resolution
                 for (const extension of supportedExtensions) {
-                    const filesInDirWithExtension = host.readDirectory(basePath, extension, exclude);

```

Comment:
We were re-reading the directory for every file extension in the list? Crazy!! Good find :smile: